src/net_processing.cpp

```diff
@@ -1948,7 +1948,7 @@ void static ProcessGetData(CNode& pfrom, Peer& peer, const CChainParams& chainpa
 
 static uint32_t GetFetchFlags(const CNode& pfrom) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
     uint32_t nFetchFlags = 0;
-    if ((pfrom.GetLocalServices() & NODE_WITNESS) && State(pfrom.GetId())->fHaveWitness) {
+    if (State(pfrom.GetId())->fHaveWitness) {
```

Comment:
I think this is buggy. Always setting `NODE_WITNESS` to our local service flags doesn't mean we will never meet peers which aren't advertising it. Of course it's unlikely there is that much of those peers deployed but otherwise I believe we'll wrongly unserialize received txn.

Reply:
Why do you think this is buggy? We'll unserialize using `MSG_WITNESS_FLAG` iff `fHaveWitness` is set to true, which happens if the peer includes `NODE_WITNESS` in its version message.

Reply:
I think we unserialize if `fAllowWitness=true`, `MSG_WITNESS_FLAG` is used by a getdata sender to require serialization of witnesses but not at unserialization itself?

That said, you're right in the sense that `fHaveWitness` implies they advertise their local services with `NODE_WITNESS`. And those ones are static so it should be good. Further, the conditional could be already reduced to `fHaveWitness` check only, a test node with `-segwitheight== std::numeric_limits<int>::max()` will have `fHaveWitness=false` ?

Reply:
> MSG_WITNESS_FLAG is used by a getdata sender to require serialization of witnesses but not at unserialization itself

Sorry, yes - I wrote "unserialize" above when I meant "serialize"

> Further, the conditional could be already reduced to fHaveWitness check only, a test node with -segwitheight== std::numeric_limits<int>::max() will have fHaveWitness=false ?

I don't think that's right. Before this PR, it's checking that both:

- we have signaled NODE_WITNESS to the peer
- the peer has signaled NODE_WITNESS to us

we can only use witness serialization on the connection when both of those conditions are true.

After this PR, we unconditionally signal NODE_WITNESS to all peers, so we only need to check that they signaled NODE_WITNESS to us (which is what is indicated by fHaveWitness).

Reply:
Gotcha, reading comment L869 in `src/net_processing.cpp` : "Note that pnode->GetLocalServices() is a reflection of the local services we were offering when the CNode object was created for this peer" makes me understood this is _our_ local services, not _their_. Renaming the method `GetAnnouncedToPeerServices` or something similar would be clearer.

But further, if our local services are static beyond initialization does it make sense to track them per-peer, the same flags should be announced to all our peers during a runtime lifecycle ?

> we can only use witness serialization on the connection when both of those conditions are true.

Does BIP144 really require double opt-in to fetch a tx and its witnesses with a getdata ? Not clear reading the BIP no more the getdata processing code.

```
            // WTX and WITNESS_TX imply we serialize with witness
            int nSendFlags = (inv.IsMsgTx() ? SERIALIZE_TRANSACTION_NO_WITNESS : 0);
            connman.PushMessage(&pfrom, msgMaker.Make(nSendFlags, NetMsgType::TX, *tx));
```

Reply:
I'm sure there are lots of improvements that could be made around how we track services (and indeed will be made as part of https://github.com/bitcoin/bitcoin/issues/19398). However, that seems orthogonal to this PR. Here, we just want to ensure that mainnet behaviour is unaffected.